app/.gitlab-ci.yml

```diff
@@ -130,15 +132,59 @@ build:media:
       - app/proto/**/*
       - app/media/**/*
 
+build:web-dev:
+  needs: ["protos"]
+  stage: build
+  script:
+    - docker pull $WEB_DEV_RELEASE_TAG || true
+    - docker build --build-arg version=$SLUG --cache-from $WEB_DEV_RELEASE_TAG -t $WEB_DEV_TAG -f app/web/dev.Dockerfile app/web/
+    - docker push $WEB_DEV_TAG
+    # creates a new docker container (docker create returns the container name), and copies the /app folder to the host
+    - mkdir -p artifacts && docker cp $(docker create $WEB_DEV_TAG):/app artifacts/web-dev
+  rules:
+    - if: $CI_COMMIT_BRANCH == $RELEASE_BRANCH
+      changes:
+      - app/**/*
+    - if: $CI_COMMIT_BRANCH != $RELEASE_BRANCH
+      changes:
+      - app/proto/**/*
+      - app/web/**/*
+  artifacts:
+    paths:
+      - artifacts/web-dev/
+
 build:web:
   needs: ["protos"]
   stage: build
   script:
     - docker pull $WEB_RELEASE_TAG || true
-    - docker build --build-arg version=$SLUG --cache-from $WEB_RELEASE_TAG -t $WEB_TAG app/web/
+    - docker build --build-arg version=$SLUG --cache-from $WEB_RELEASE_TAG -t $WEB_TAG -f app/web/prod.Dockerfile app/web/
     - docker push $WEB_TAG
     # creates a new docker container (docker create returns the container name), and copies the /app folder to the host
-    - mkdir -p artifacts && docker cp $(docker create $WEB_TAG):/app/build artifacts/web
+    - mkdir -p artifacts && docker cp $(docker create $WEB_TAG):/app artifacts/web
+  rules:
+    - if: $CI_COMMIT_BRANCH == $RELEASE_BRANCH
+      changes:
+      - app/**/*
+    - if: $CI_COMMIT_BRANCH != $RELEASE_BRANCH
+      changes:
+      - app/proto/**/*
+      - app/web/**/*
+  artifacts:
+    paths:
+      - artifacts/web/
+
+build:web-next:
+  needs: ["protos"]
+  stage: build
+  script:
+    - docker pull $WEB_NEXT_RELEASE_TAG || true
+    # this is a bit hacky, but I don't see any other good way to get the preview env vars into the container
+    - cp app/web/.env.preview app/web/.env.production
```

Comment:
You can do this in the dockerfile. This is how the `nginx` containers picks production or preview. I think it's cleaner to do it there.

Short version:

1. In dockerfile

```dockerfile
# 'production' or 'preview'
ARG environment=production

COPY .env.${environment} .env
```

2.
Add docker build arg: `--build-arg environment=preview`